MP-5134 Table Feature Editable (#257)
spryker/spryker-docs: docs/marketplace/dev/front-end/table-design/table-features/table-feature-editable.md

```html
<spy-table [config]="{
  dataSource: { ... },
  columns: [ ... ],
  editable: {
    columns: [ ... ],
    create: { ... },
    update: { ... },
    disableRowKey: 'col',
  },                                                                                           
}">
</spy-table>
```

Let's take a closer look at all the possible options.

`columns`:  
  - `id` - a cell `id`.  
  - `type` - a cell `type`.  
  - `typeOptions`:  
    - `value` - will set default value to the cell of the newly added row.  

`create`:  
  - `addButon` - an object with `title` and `icon` for the `Add button`.  
  - `cancelButon` - an object with `title` and `icon` for the `Cancel button`.  
  - `disableForCols` - an array with cell `id's` to be disabled.  
  - `formInputName` - will create `input[type=hidden]` element with the specific name.  
  - `initialData` - initials data for cells and object with errors for rows and cells.  
  
`update`:
  - `url` - request url.  
  - `saveButon` - an object with `title` and `icon` for the `Save button` (displays in the `update` popup).  
  - `cancelButon` - an object with `title` and `icon` for the `Cancel button` (displays in the `update` popup).  

>   - `disableForCols` - an array with cell `id's` to be disabled.  `

Reviewer: ```suggestion
  - `disableForCols` - an array with cell `ids` to be disabled.  `
```

  update?: TableEditableConfigUpdate;
  disableRowKey?: string;
}

export interface TableEditableConfigCreate {
  formInputName: string;
  initialData?: TableEditableConfigCreateData;
  addButton?: TableEditableConfigButton;
  cancelButton?: TableEditableConfigButton;
  disableForCols?: string[];
}

export interface TableEditableConfigUpdate {
  url: TableEditableConfigUrl;
  saveButton?: TableEditableConfigButton;
  cancelButton?: TableEditableConfigButton;
  disableForCols?: string[];
}

export interface TableEditableConfigCreateData {
  data: TableDataRow[];
  errors?: TableEditableConfigDataErrors;
}

export interface TableEditableConfigDataErrorsFields {
  rowError?: string;
  columnErrors?: { [columnId: string]: string | undefined };
}

export interface TableEditableConfigDataErrors {
  [rowIdx: string]: TableEditableConfigDataErrorsFields;
}

export interface TableEditableConfigUrlObject {
  url: string;
  method?: string;
}

export type TableEditableConfigUrl = string | TableEditableConfigUrlObject;

export interface TableEditableConfigButtonIcon {
  icon: string;
}

export interface TableEditableConfigButtonText
  extends Partial<TableEditableConfigButtonIcon> {

>   title: string;
> }

Reviewer: ```suggestion
  title: string;
  size: ButtonSize;
  shape: ButtonShape;
  variant: ButtonVariant;
  type: ButtonType;
}
```

description: This article provides details about the Table Feature Editable component in the Components Library.
template: concept-topic-template
---

This article provides details about the Table Feature Editable component in the Components Library.

## Overview

Table Feature Editable is a feature of the Table Component that allows editing/adding rows of the table.
Below you can see base config of the Editable feature, that contains next options: 

Feature Configuration:

`columns` - an array with config for each editable column.  
`create` - an object with config for a new rows.  
`update` - an object with config for an existing rows.  
`disableRowKey` - will disable row that contains mentioned column `id` (see example).  

```html
<spy-table [config]="{
  dataSource: { ... },
  columns: [ ... ],
  editable: {
    columns: [ ... ],
    create: { ... },
    update: { ... },
    disableRowKey: 'col',
  },                                                                                           
}">
</spy-table>
```

Let's take a closer look at all the possible options.

`columns`:  
  - `id` - a cell `id`.  
  - `type` - a cell `type`.  
  - `typeOptions`:  

>     - `value` - will set default value to the cell of the newly added row.  

Reviewer 2: Please, mention that you can use every `typeOptions` variant depended on table column type (with link to it)

`create`:  

>   - `addButon` - an object with `title` and `icon` for the `Add button`.  

Reviewer 2: It was updated recently, please fix it in this doc too

      addButton: {
        title: 'Add price',
        icon: 'icon',
      },
      cancelButton: {
        title: 'Cancel',
        icon: 'icon',
      },
      disableForCols: ['col2'],
      formInputName: 'form-input-name',
      initialData: {
        data: [
          { col1: 'value', col2: 'value' },
          { col1: 'value' },
        ],
        errors: {
          0: {
            rowError: 'message',
            columnErrors: {
              col3: 'errorMessage errorMessage errorMessage',
            },
          },
        },
      },
    },
    update: {
      url: '/table-update-cell',
      saveButton: {
        title: 'Save',
        icon: 'icon',
      },
      cancelButton: {
        title: 'Cancel',
        icon: 'icon',
      },
      disableForCols: ['col2'],
    },
    disableRowKey: 'col1',
  },                                                                                           
}">
</spy-table>
```

## Feature Registration

```ts
@NgModule({
  imports: [
    TableModule.forRoot(),
    TableModule.withFeatures({
      editable: () => 
        import('@spryker/table.feature.editable').then(
          (m) => m.TableEditableFeatureModule,
        ),   

>     }),

Reviewer 2: Also show example without dynamic registration, just feature tag inside table

Author: added to the tech debt https://spryker.atlassian.net/browse/MP-5886

  ],
})
export class RootModule {}
```

## Interfaces

Below you can find interfaces for Table Feature Editable.

```ts

> export interface TableEditableColumn extends TableColumn {

Reviewer 2: Please, add interface to show that we extend default `TableConfig` with this feature

```
declare module '@spryker/table' {
  interface TableConfig {
    editable?: TableEditableConfig;
  }
}
```

Author: added to the tech debt https://spryker.atlassian.net/browse/MP-5886

Below you can see base config of the Editable feature, that contains next options: 

Feature Configuration:

`columns` - an array with config for each editable column.  
`create` - an object with config for a new rows.  
`update` - an object with config for an existing rows.  
`disableRowKey` - will disable row that contains mentioned column `id` (see example).  

```html
<spy-table [config]="{
  dataSource: { ... },
  columns: [ ... ],
  editable: {
    columns: [ ... ],
    create: { ... },
    update: { ... },
    disableRowKey: 'col',
  },                                                                                           
}">
</spy-table>
```

Let's take a closer look at all the possible options.

`columns`:  
  - `id` - a cell `id`.  
  - `type` - a cell `type`.  
  - `typeOptions` (see more about [Column Type](/docs/marketplace/dev/front-end/table-design/table-column-types/) options):  
    - `value` - will set default value to the cell of the newly added row.  

`create`:  
  - `addButon` - an object with the `Add button` configuration like `title`, `icon`, `size`, etc..  

>   - `cancelButon` - an object with `title` and `icon` for the `Cancel button`.  

Reviewer: ```suggestion
  - `cancelButon` - an object with the `Cancel button` configuration like `title`, `icon`, `size`, etc.
```

  - `disableForCols` - an array with cell `ids` to be disabled.  
  - `formInputName` - will create `input[type=hidden]` element with the specific name.  
  - `initialData` - initials data for cells and object with errors for rows and cells.  
  
`update`:
  - `url` - request url.  

>   - `saveButon` - an object with `title` and `icon` for the `Save button` (displays in the `update` popup).  
>   - `cancelButon` - an object with `title` and `icon` for the `Cancel button` (displays in the `update` popup).  

Reviewer: ```suggestion
  - `saveButon` - an object with the `Save button` configuration like `title`, `icon`, `size`, etc (displays in the `update` popup).  
  - `cancelButon` - an object with the `Cancel button` configuration like `title`, `icon`, `size`, etc (displays in the `update` popup).  
```

export class RootModule {}
```

```html

> // Via HTML

Reviewer 2: Mention, that for this approach we need add TableEditableFeatureModule

```
@NgModule({
  imports: [
    TableModule.forRoot(),
    TableEditableFeatureModule,
  ],
})
export class RootModule {}
```
